lib/Sema/PreCheckExpr.cpp

```diff
@@ -238,6 +238,20 @@ static bool diagnoseRangeOperatorMisspell(DiagnosticEngine &Diags,
   return false;
 }
 
+static bool diagnoseUnexistingOperators(DiagnosticEngine &Diags,
+                                          UnresolvedDeclRefExpr *UDRE) {
+  auto name = UDRE->getName().getBaseIdentifier();
+  if (!name.isOperator())
+    return false;
+
+  if (name.str() == "**") {
+    Diags.diagnose(UDRE->getLoc(), diag::unexisting_power_operator).highlight(UDRE->getSourceRange());
+    return true;
+  }
```

Comment:
Couple of minor notes for improvement here: (based on combined feedback from Joe and Becca)

`pow` is actually defined in the C library, and Foundation re-exports it, so maybe suggesting Foundation is not the right thing to do. Instead, we can do an unqualified lookup using `getModuleScopeContext()` (to get the scope corresponding to the current file) and the `TypeChecker::lookupUnqualified` method, to check if there is a `pow` in scope. If `pow` is in scope, we can issue an error like `no operator '**' is defined; did you mean 'pow(_:_:)'?`.

Otherwise, we don't issue the special error.

For test the diagnostic, I think you should be able to test this using either (1) the fake SDK that we have `%clang-importer-sdk` and have that provide a `pow` function OR (2) creating a small test case where one module exports `pow` and another module that imports the first one tries to use `**` without defining it.

Reply:
Thanks! I'm not sure if this can be merged since @xwu already created a proposal, but I changed it to lookup for `pow` as you suggested :)

Reply:
> I'm not sure if this can be merged [..]

I think your PR is already in good shape and provides a nice improvement, so we should merge it. From a cursory reading, the discussion on the forums seems to be a bit animated (different people are pointing out that different behavior is desirable), and then it will also need to be scheduled for review + core team approval. I'm guessing the proposal should be accepted, but I think we should merge this in case that doesn't pan out.

Reply:
I agree; I certainly have no objections to trying to improve the diagnostics here. I like @varungandhi-apple's suggestions above.